src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -763,6 +763,78 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                     preserveSig);
             }
         }
+
+        private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is object);
+
+            if (MethodKind != MethodKind.Ordinary)
+            {
+                // Ignore cases where there will already be a warning due to the AttributeTargets.Method usage
```

Comment:
>Ignore cases where there will already be a warning due to the AttributeTargets.Method usage [](start = 19, length = 91)

What is left? We shouldn't assume any specific usage specified on the attribute and should test all possible scenarios.

Reply:
I can do that. I asked this question up front and @RikkiGibson thought I should not produce diagnostics in these situations.

Reply:
> I can do that. I asked this question up front and Rikki Gibson thought I should not produce diagnostics in these situations.

I think we will not get here for an invalid target, so there is no risk to produce a duplicate diagnostics (should be easy to verify with a test). Other than that, we should not be making any assumptions that APIs have specific shape, etc. Everything should be verified. We do not trust and we do not assume. 

---
In reply to: [418182673](https://github.com/dotnet/roslyn/pull/43301#discussion_r418182673) [](ancestors = 418182673)

Reply:
Given that get here only for methods with the specific attributes. What those methods could be?

---
In reply to: [418187450](https://github.com/dotnet/roslyn/pull/43301#discussion_r418187450) [](ancestors = 418187450,418182673)

Reply:
> I think we will not get here for an invalid target, so there is no risk to produce a duplicate diagnostics (should be easy to verify with a test)

I did add that test and that is why I had to add this check. We do get here for an invalid target. Constructors and static constructors both decode the attribute even when the attribute's usage makes it invalid on constructors.

> Other than that, we should not be making any assumptions that APIs have specific shape, etc. Everything should be verified. We do not trust and we do not assume.

This has been my mindset.

> Given that we get here only for methods with the specific attributes (ignoring attribute target). What those methods could be?

We get here for finalizers, operators, accessors, and local functions (permitted by AttributeTargets.Method) and we get here for constructors and static constructors (when the attribute is misapplied).

Reply:
> We get here for finalizers, operators, accessors, and local functions (permitted by AttributeTargets.Method) and we get here for constructors and static constructors (when the attribute is misapplied).

Given how big is the list, it might make sense to stick with the short form of the message "an ordinary member method".

---
In reply to: [418200095](https://github.com/dotnet/roslyn/pull/43301#discussion_r418200095) [](ancestors = 418200095)

Reply:
> I did add that test and that is why I had to add this check. We do get here for an invalid target. Constructors and static constructors both decode the attribute even when the attribute's usage makes it invalid on constructors.

This doesn't match my experience. For this code ```DecodeWellKnownAttributeAppliedToMethod``` is never called:
```
[System.AttributeUsage(System.AttributeTargets.Method)]
class Test : System.Attribute
{
    [Test]
    Test()
    {

    }

    [Test]
    static Test()
    {

    }
}
```

However, if change the attribute usage to ```[System.AttributeUsage(System.AttributeTargets.Constructor)]```, it is called twice.
What am I missing?

---
In reply to: [418233735](https://github.com/dotnet/roslyn/pull/43301#discussion_r418233735) [](ancestors = 418233735,418200095)

Reply:
The problem lies in my test which is missing AttributeUsage. I'm sorry for the confusion.

Okay, so what are the changes I should be making?

- I should insert the word "member" into the current diagnostic to obtain your suggestion, "an ordinary member method"
- Are you asking for a change to these constructor checks? If I keep the constructor checks, the result will be consistent with what would happen if the attribute was applied to non-method targets like classes: no diagnostic.

Reply:
> I should insert the word "member" into the current diagnostic to obtain your suggestion, "an ordinary member method"

I would do this.

> Are you asking for a change to these constructor checks? If I keep the constructor checks, the result will be consistent with what would happen if the attribute was applied to non-method targets like classes: no diagnostic.

I do not think we need any extra checks, as long as the member is not ordinary, we complain. It is quite possible that I am missing something. In this case, share a scenario and observed behavior with an explanation why do you think it is "not good" and with your understanding why compiler behaves this way. 

---
In reply to: [418333600](https://github.com/dotnet/roslyn/pull/43301#discussion_r418333600) [](ancestors = 418333600)

Reply:
@AlekseyTs This is the scenario:

```cs
class C
{
    // No error because AttributeUsage should have prevented this and we don't care if it is misdefined.
    [ModuleInitializer]
    public C SomeProperty { get; }

    // AttributeUsage should have prevented this and we don't care if it is misdefined.
    // Wait, we're going to report an error anyway:
    // ❌ Module initializer must be an ordinary member method.
    [ModuleInitializer]
    public C() { }
}

namespace System.Runtime.CompilerServices
{
    class ModuleInitializerAttribute : Attribute { }
}
```

Reply:
> AttributeUsage should have prevented this

Why do you think this should happen? I don't see any AttributeUsage  restriction on the attribute, we do not assume any restrictions on the usage, we behave based on the actual restrictions:
```
namespace System.Runtime.CompilerServices
{
    class ModuleInitializerAttribute : Attribute { }
}

```

> Wait, we're going to report an error 

Given the code, this is exactly the behavior I expect. 

---
In reply to: [418346434](https://github.com/dotnet/roslyn/pull/43301#discussion_r418346434) [](ancestors = 418346434)

Reply:
> > AttributeUsage should have prevented this
>
> Why do you think this should happen? I don't see any AttributeUsage restriction on the attribute, we do not assume any restrictions on the usage, we behave based on the actual restrictions

What I'm saying should have happened is that the person who defined the attribute should have used the AttributeUsage from the spec.

Reply:
> Given the code, this is exactly the behavior I expect.

Sounds good! I'll remove the constructor check.

Reply:
> What I'm saying should have happened is that the person who defined the attribute should have used the AttributeUsage from the spec.

I think this was mentioned several times, we are not supposed to make assumptions like that. And we shouldn't take dependency on someone else do something in a particular way.

---
In reply to: [418785472](https://github.com/dotnet/roslyn/pull/43301#discussion_r418785472) [](ancestors = 418785472)